File: 940-fruit-into-baskets/fruit-into-baskets.cpp
class Solution {
public:
    int totalFruit(vector<int>& fruits) {

        int n = fruits.size();
        int i = 0;
        int first = fruits[0];

        int k = 0;
        while (k < fruits.size() && fruits[k] == first) {
            k++;
        }

        if (k == n)
            return n;

        int y = k;
        int next = k;
        int second = fruits[k];

        while (y < fruits.size() && fruits[y] == fruits[k]) {
            y++;
        }

        if(y == n) return n;

        int j = y;
        int ans = j;


        while (j < fruits.size()) {

            // cout << first << " " << second << endl;

            if (fruits[j] == first || fruits[j] == second) {
                next = j;
                int k = j;
                while (k < fruits.size() && fruits[k] == fruits[j]) {
                    k++;
                }
                j = k - 1;
            } else {
                i = next;
                next = j;
                int k = j;
                while (k < fruits.size() && fruits[k] == fruits[j]) {
                    k++;
                }
                j = k - 1;
                first = fruits[i];
                second = fruits[j];
            }
            ans = max(ans, j - i + 1);
            j++;
        }

        return ans;
    }
};

File: 940-fruit-into-baskets/fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {

        HashMap<Integer, Integer> map = new HashMap<>()  ;
        int maxLen = 0 ; 
        int start = 0 ; 

        for(int i = 0; i < fruits.length ; i++){
            
            int n = fruits[i] ;

            map.put(n , map.getOrDefault(n,0) + 1);

            if(map.size() > 2){

                int k = fruits[start] ;

                map.put(k , map.get(k) - 1);

                if(map.get(k) == 0){
                    map.remove(k);
                }

                start++; 
            }


            if(map.size() == 2){
                maxLen = Math.max(maxLen , i - start + 1);
            }

        }

        if(map.size() < 2)return fruits.length ; 
        return maxLen ; 
        
    }
}